pkg/controllers/secretstore/common.go

```diff
@@ -0,0 +1,111 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package secretstore
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esapi "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	"github.com/external-secrets/external-secrets/pkg/provider/schema"
+)
+
+const (
+	errStoreProvider       = "could not get store provider: %w"
+	errStoreClient         = "could not get provider client: %w"
+	errValidationFailed    = "could not validate provider: %w"
+	errPatchStatus         = "unable to patch status: %w"
+	errUnableCreateClient  = "unable to create client"
+	errUnableValidateStore = "unable to validate store"
+	errUnableGetProvider   = "unable to get store provider"
+
+	msgStoreValidated = "store validated"
+)
+
+func reconcile(ctx context.Context, req ctrl.Request, ss esapi.GenericStore,
+	cl client.Client, log logr.Logger, controllerClass string, recorder record.EventRecorder) (ctrl.Result, error) {
+	if !ShouldProcessStore(ss, controllerClass) {
+		log.V(1).Info("skip store")
+		return ctrl.Result{}, nil
+	}
+
+	// patch status when done processing
+	p := client.MergeFrom(ss.Copy())
+	defer func() {
+		err := cl.Status().Patch(ctx, ss, p)
+		if err != nil {
+			log.Error(err, errPatchStatus)
+		}
+	}()
+
+	// validateStore modifies the store conditions
+	// we have to patch the status
+	log.V(1).Info("validating")
+	err := validateStore(ctx, req.Namespace, ss, cl, recorder)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	recorder.Event(ss, v1.EventTypeNormal, esapi.ReasonStoreValid, msgStoreValidated)
+	cond := NewSecretStoreCondition(esapi.SecretStoreReady, v1.ConditionTrue, esapi.ReasonStoreValid, msgStoreValidated)
+	SetExternalSecretCondition(ss, *cond)
+
+	return ctrl.Result{}, err
```

Comment:
Wouldn't it be better add a `RequeueAfter` with a constant value (for now, at least), and catch day 2 operational errors on the events?

Reply:
Requeue | RequeueAfter | Error | Result
-- | -- | -- | --
any | any | !nil | Requeue with rate limiting.
true | 0 | nil | Requeue with rate limiting.
any | >0 | nil | Requeue after specified RequeueAfter.
false | 0 | nil | Do not requeue.

I think Gustavo is right, we can requeue: true here and let it Requeue with rate limiting.

Reply:
The [DefaultRatelimiter](https://github.com/kubernetes/client-go/blob/v0.23.3/util/workqueue/default_rate_limiters.go#L39-L45) delays between 1ms and 1000s. When using `Requeue=true` we will end up with a delay of 1000s in the long run because the "failure" won't be reset (`Queue.Forget()` [here](https://github.com/kubernetes-sigs/controller-runtime/blob/4d6d8ef2bb3191239ccc958691b790854b83faa3/pkg/internal/controller/controller.go#L323-L332)). 

I think we should stick with a static interval (configurable with flag, default to 5m ?).